### PR TITLE
[pip][design] PIP 289: Secure Pulsar Connector Configuration

### PR DESCRIPTION
**Pull request by michaeljmarshall**
Fixes: https://github.com/apache/pulsar/issues/20862

### Motivation

PIP for improving connector configuration security. See content for details.

### Modifications

Add the design doc for PIP 289.

Current implementation PRs:
* https://github.com/apache/pulsar/pull/20901
* https://github.com/apache/pulsar/pull/20902

### Documentation

- [x] `doc`

### CODE REVIEW

File: pip/pip-289.md [hunk 1]
```diff
@@ -0,0 +1,256 @@
...
+interface SecretsProvider {
+    /**
+     * If the passed value is formatted as a reference to a secret, as defined by the implementation, return the
+     * referenced secret. If the value is not formatted as a secret reference or the referenced secret does not exist,
+     * return null.
+     *
+     * @param value a config value that may be formatted as a reference to a secret
+     * @return the materialized secret. Otherwise, null.
+     */
+    default String interpolateSecretForValue(String value) {
```
Comment (shibd) [inline pip/pip-289.md:118]: Sorry, I'm not clear about this method.

Who will call it? Can you provide examples?

File: pip/pip-289.md [hunk 2]
```diff
@@ -0,0 +1,256 @@
...
+
+Add `interpolateSecretsIntoConfigMap` to the `WorkerConfig`. This configuration will allow operators to determine whether users can interpolate secrets into configurations. Here is the potential documentation for this flag:
+
+```java
+class WorkerConfig {
+    @FieldContext(
+            category = CATEGORY_CONNECTORS,
+            doc = "Whether to interpolate secrets into the connector's configuration. Only affects Sinks and Sources."
+                    + " Applies to all connectors deployed by this function worker."
+                    + " When true, the SecretsProvider will recursively interpolate secrets into the connector's "
```
Comment (shibd) [inline pip/pip-289.md:160]: What would we do with this configuration? Can you show you an overview of the implementation code?

> When true, the SecretsProvider will recursively interpolate secrets into the connector's .

Is the `SecretsProvider` implemented by the user himself?

**Comment by shibd** [comment]
On the three problems you mentioned:

- Elasticsearch Config(Nested configurations): We can improve `IOConfigUtils` to support it. Right?
- RabbitMQ Config(Not using IOConfigUtils to load config): We can find and fix connectors for these same problems. Right?
- Kafka Config(Use kafka properties map): 
  - Perhaps we should refine these keys to create configurations separately
  - Or we can enhance annotations to support set sensitive key names, such as:
```java
    @FieldDoc(
            defaultValue = "",
            help = "Config properties to pass to the kafka connector."
            sensitiveNames="password-1, password-2"
    )
    private Map<String, String> kafkaConnectorConfigProperties;
```

File: pip/pip-289.md (hunk 1 above)
Comment (michaeljmarshall) [reply pip/pip-289.md:118]: I provided an implementation here: https://github.com/apache/pulsar/pull/20901. This method will be called by Pulsar.

**Comment by michaeljmarshall** [comment]
> Elasticsearch Config(Nested configurations): We can improve `IOConfigUtils` to support it. Right?

It's possible, but it would likely require a change/expansion to the way the `FieldDoc` annotation works. However, fixing it for Elasticsearch does not solve the general problem.

> RabbitMQ Config(Not using IOConfigUtils to load config): We can find and fix connectors for these same problems. Right?

I already proposed this as part of the PIP here: https://github.com/apache/pulsar/pull/20902.

> * Kafka Config(Use kafka properties map):
>       
>       * Perhaps we should refine these keys to create configurations separately
>       * Or we can enhance annotations to support set sensitive key names, such as:
> 
> 
> ```java
>     @FieldDoc(
>             defaultValue = "",
>             help = "Config properties to pass to the kafka connector."
>             sensitiveNames="password-1, password-2"
>     )
>     private Map<String, String> kafkaConnectorConfigProperties;
> ```

The Kafka Config is the real reason for this PIP. We need arbitrary interpolation because the Kafka Config works as a wrapper. For example, it is possible to configure it with any connector that works with Kafka. See https://github.com/datastax/pulsar-3rdparty-connector for many third party connectors that are simply plugged into the Kafka Adapter Connector without any additional work. Since it is a wrapper, it is not possible to annotate the relevant fields with a custom Pulsar annotation (`FieldDoc`). Instead, I propose we find a more general solution.

File: pip/pip-289.md (hunk 2 above)
Comment (michaeljmarshall) [reply pip/pip-289.md:160]: Please see https://github.com/apache/pulsar/pull/20901. The `SecretsProvider` is somewhat configurable (though the code is not very straightforward).

**Comment by shibd** [comment]
Hi,  @michaeljmarshall. Thanks for your explanation. I got it.

I think we can avoid adding a new config: `interpolate_secrets_into_config_map`.

What's the problem with us calling the method by default?
```
    private static void interpolateSecretsIntoConfigs(SecretsProvider secretsProvider, Map<String, Object> configs)
```

Only when the user uses placeholders will it merge secrets to the config map, right?
```
      "username": "${username}",
      "password": "${password}",
```

This way we just need to explain to the user that we provide a new way to set secrets.

**Comment by michaeljmarshall** [comment]
> What's the problem with us calling the method by default?

Great question, thanks @shibd. I wrote this text in the PIP explaining why I didn't want to enable it by default:

> The primary security consideration is whether there is any risk in giving users a way to interpolate environment variables into their connector. Note that this kind of feature led to Log4Shell. Unlike Log4Shell, the risk is negligible when running in a containerized environment, like Kubernetes. Further, this feature will be disabled by default, so users can evaluate the security risks on their own.

In taking a closer look, the proposed change will only affect k8s based deployments, which probably means we can enable it by default. However, the only remaining risk could be to users that have created their own extensions and are using these classes in unknown ways. Given your comments, I think we should remove the configuration option and always enable it. Let me know what you think, thanks.

**Review by eolivelli — APPROVED** [review]
Lgtm 

+1 (binding)

**Comment by michaeljmarshall** [comment]
@shibd - I updated the text of the PIP. PTAL

**Review by shibd — APPROVED** [review]
Nice work!

**Pull request merged**